README.md
# Oracle compiler for Atom

[![apm](https://img.shields.io/apm/l/build-oracle.svg?style=flat-square)](https://atom.io/packages/build-oracle)
[![apm](https://img.shields.io/apm/v/build-oracle.svg?style=flat-square)](https://atom.io/packages/build-oracle)
[![apm](https://img.shields.io/apm/dm/build-oracle.svg?style=flat-square)](https://atom.io/packages/build-oracle)

Uses the [atom-build](https://atom.io/packages/build) package to execute Oracle compilations in the Atom editor.

This package requires [atom-build](https://atom.io/packages/build) to be installed.

## Prerequisites

To be able to compile your code, you need either `SQL*Plus` or `SQLcl` available on your system - preferably in your PATH. `SQL*Plus` requires more steps to get set up (such as installing the instant client), but you will get much faster compile times so that is the preferred interpreter (`SQLcl` is Java based, so launching the JVM each time adds to the compile time).

### SQL*Plus set up

#### OS X

You may face issues depending on how you set up the instant client. The [typical guide](https://docs.oracle.com/cd/E11882_01/install.112/e38228/inst_task.htm#BABHEBIG) involves setting an environment variable `DYLD_LIBRARY_PATH`. Unforunately, in OS X El Capitan, this variable is not passed in as an environment variable into Atom (this is a part of security measure known as System Integrity Protection). You can also see that this is not output if you run the `env` command on the terminal.

If you followed those general steps, you will likely be getting this error (or very similar):

```
dyld: Library not loaded: /ade/dosulliv_sqlplus_mac/oracle/sqlplus/lib/libsqlplus.dylib
  Referenced from: /opt/Oracle/instantclient_11_2/sqlplus
  Reason: image not found
Trace/BPT trap: 5
```

Two ways around this.

* In the settings for this plugin, set the path where your client files are (copy what you have for your existing DYLD_LIBRARY_PATH).

![](https://cloud.githubusercontent.com/assets/1747643/17292725/1f61b8a2-582f-11e6-940f-b9c7ae7ff42e.png)

* Or, instead of setting `DYLD_LIBRARY_PATH`, you can make all the libaries available in a location that `SQL*Plus` knows where to look for them. Two common locations are `~/lib` or `/usr/local/lib`. So, assuming you placed the instant client at `/opt/Oracle/instantclient_11_2`, run the following:

```bash
sudo mkdir -p /usr/local/lib
sudo ln -s /opt/Oracle/instantclient_11_2/libclntsh.dylib.11.1 /usr/local/lib/libclntsh.dylib.11.1
sudo ln -s /opt/Oracle/instantclient_11_2/libnnz11.dylib /usr/local/lib/libnnz11.dylib
sudo ln -s /opt/Oracle/instantclient_11_2/libociei.dylib /usr/local/lib/libociei.dylib
sudo ln -s /opt/Oracle/instantclient_11_2/libsqlplus.dylib /usr/local/lib/libsqlplus.dylib
sudo ln -s /opt/Oracle/instantclient_11_2/libsqlplusic.dylib /usr/local/lib/libsqlplusic.dylib
```

Ensure your `PATH` is set up with the instant client directory on it so that `sqlplus` is on it. In your `~/.bash_profile` add the line:

```bash
export PATH=/opt/Oracle/instantclient_11_2:$PATH
```

I went with `/usr/local/lib` rather than `~/lib` so to not fill up my home directory. Aside from that difference, this is much as what is described on the [download page](http://www.oracle.com/technetwork/topics/intel-macsoft-096467.html) of the instant client (down the bottom of the page).

If you previously set up `DYLD_LIBRARY_PATH` you can unset that and you should be good to go.

#### Ubuntu/Linux

To be able to run SQL*Plus from Atom, Atom needs to have all the relevant enviornment variables set up. I followed the installation from this [guide](https://help.ubuntu.com/community/Oracle%20Instant%20Client). The guide mentions to set environment variables in a system wide location or `~/.bash_profile`. The guide suggests to install it in a system wide location since you typically only have one Oracle client installed per system. That is a good approach also for Atom, as it will ensure Atom picks up the relevant environment variables. I found that environment variables set in `~/.bash_profile` are not picked up by the (GUI) terminal, and upon further digging I found that `~/.bash_profile` is only picked up when launching a virtual console - see this [post](http://askubuntu.com/questions/121073/why-bash-profile-is-not-getting-sourced-when-opening-a-terminal). The other location you may be tempted to use is `~/.bashrc`, however this is also not a good place as Atom doesn't seem to pick up the environment set up in there, see this [comment](https://github.com/joefitzgerald/go-plus/issues/386#issuecomment-199359955).

So, to sum it up. Either place your Oracle variables in a script at `/etc/profile.d/oracle.sh` (globally set), or place them in `~/.profile` which will get picked up by Atom.

### SQLcl set up

#### Windows

An issue you may face on Windows is if you try and compile code containing unicode characters, they won't persist when compiling with SQLcl. The reason is that the JVM doesn't default to UTF-8 encoding on Windows.

To get around this, ensure you point to sql.bat (instead of sql.exe) and also that you have checked "Enforce UTF-8" in the package settings, similar to:

![screenshot from 2016-08-28 17-49-57](https://cloud.githubusercontent.com/assets/1747643/18032450/6bcbdfbe-6d48-11e6-87f0-a471ff298301.png)

When `Enforce UTF-8 encoding` is on, the environment variable `JAVA_OPTS` will be set to `-Dfile.encoding=UTF-8`. This is also discussed on Stack overflow: http://stackoverflow.com/questions/361975/setting-the-default-java-character-encoding

## Installation

Install through apm or in Atom itself, where the name of the package is `build-oracle`.

```
apm install build-oracle
```

Once installed, you will want to access the settings and set the relevant path for your SQL interpreter. That will be either SQLcl or SQL*Plus. For me, I have `sqlplus` in my `path` so no change is necessary (there is an issue where PATH may not be correctly set in Atom, in that case you should set the full path of `sqlplus`). If I wanted to point it to the binary for `SQLcl`, which isn't in my `path`, I would set the configuration to `/opt/sqlcl/bin/sql`.

![](https://cloud.githubusercontent.com/assets/1747643/11413201/79fcaa10-943a-11e5-881f-1715ef163e29.png)

## Usage instructions

In your project root directory, add a JSON file named `.atom-build-oracle.json`, where you can define build targets for your project. This contents of the file expects an array of objects, which supports the following parameters:

|Property name  | Description             | Example |
|---            |---                      | ---     |
| targetName    | A descriptive name of the build target. This is so you know to which environment you'll be compiling to.       | hr: dev|
| connectString | The connect string you would used to connect to `sqlplus` on the command line. <br>**note 1:** You *must* specify a password at this stage. <br>**note 2:** `connectString` is the preferred field to use. | - hr/hr@//server:port/sid  <br />- hr/hr@XE1 |
| host | The server of the database       | example.com |
| port | The database port                | 1521 (default) |
| sid  | The database SID or service name | XE |
| user | The schema name you connect to   | hr |
| password | The password of the schema   | oracle |

note: `connectString` is built from the other server details if you ommit that setting.

Examples:

Connect string:

Suppose I have the following TNS entry defined:

```
XE1 =
    (DESCRIPTION =
        (ADDRESS_LIST =
            (ADDRESS = (PROTOCOL = TCP)(HOST = 192.168.56.101)(PORT = 1521))

        )
        (CONNECT_DATA =
            (SERVICE_NAME = XE)
        )
    )
```

My build file becomes:

```json
[
    {
        "targetName" : "hr: dev",
        "connectString" : "hr/hr@XE1"
    }
]
```

Individual details:
```json
[
    {
        "targetName" : "hr: dev",
        "host" : "example.com",
        "sid" : "xe",
        "port" : 1521,
        "user" : "hr",
        "password" : "hr"
    }
]
```

Where you can specify any number of build targets.
If your database server uses service names instead of SID, you can enter the service name in the property "sid" nonetheless.

Since this file contains sensitive information (password) you will likely also want to add an entry to your `.gitignore` file so this is not published.

```conf
.atom-build-oracle.json
```

It's worth noting, that to get the build system to recognise the above configuration, you will need to reload atom (only the first time). This can be done in the menu `View->Reload` or with the keyboard shortcut `ctrl+alt+r`.

This plugin is an extension to the `build` system, providing the necessary scripts to compile Oracle code. The build triggers are determined as part of the over-arching `build` package. To build the active file is done with the keyboard shortcut `ctrl+alt+b` or `cmd+alt+b` or `f9`. By default, this will be the first entry in your build config file. The active build target can be switched with `f7`, which will also trigger a build on the active file:

![](https://cloud.githubusercontent.com/assets/1747643/15595301/2eeb787e-2401-11e6-9b89-f138d261c122.png)

The build panel (with build output) can be toggled with `f8`.


CHANGELOG.md
# Changelog

## 1.6.1

* Resolve bug with installing dependencies

## 1.6.0

* Add support for compiling an untitled file

## 1.5.2

* Enable `sqlblanklines`

## 1.5.1

* Add default port (Issue [#32](https://github.com/tschf/atom-build-oracle/issues/32))
* Call sqlplus with the -L argument

## 1.5.0

* Add support for `connectString` (and with it, tnsnames.ora)

## 1.4.0

* Add DYLD_LIBRARY_PATH setting for OS X

## 1.3.0

* Set up dependencies to install `atom-build`
* Remove redundant installation instructions

## 1.0.0

* Publish to APM
* Add config for `NLS_LANG`
* Add error matching to report on failed builds
* Support relative paths

## 0.1.0

Initial version. Add core functionality to compile code.


exp/backgroundRunner.sh
#!/bin/bash
# DEFINE EXIT CODES
INTERPRETER_NOT_FOUND=1
TIMEOUT=2

PKG_DIR=$1
SQL_INTERPRETER=$2

TMP_DIR=${PKG_DIR}/tmp
FIFO_PATH=${TMP_DIR}/sqlconsole
SPOOL_TO=${TMP_DIR}/spooledSql.txt
DEBUG_FILE=${TMP_DIR}/fifoStartup.log
DEBUG_ENABLED=N
INTERPRETER_PATH=$(which ${SQL_INTERPRETER})

function makeFifo {
    debug "Creating new FIFO at ${FIFO_PATH}"
    rm -f ${FIFO_PATH}
    mkdir -p ${TMP_DIR}
    mkfifo ${FIFO_PATH}
    (tail -f ${FIFO_PATH} | ${INTERPRETER_PATH} /nolog &)
    debug "Process started with pid $!"
}

function debug {
    MSG=$1
    if [[ "${DEBUG_ENABLED}" -eq "Y" ]]; then
        echo ${MSG} >> ${DEBUG_FILE}
    fi

}

if [[ ! -x ${INTERPRETER_PATH} ]]; then
    echo "${SQL_INTERPRETER} either can not be found on the system, or is not executable" >&2
    exit ${INTERPRETER_NOT_FOUND}
fi

> ${DEBUG_FILE}

if [[ -e ${FIFO_PATH} ]]; then
#The fifo already exists, so lets check if sqlcl is running behind it
#We can do so by spooling a file to see if it gets created
    rm -f ${SPOOL_TO}
    debug "FIFO already exists on the file system"
    echo "SPOOL ${SPOOL_TO}" > ${FIFO_PATH} &
    SPOOL_PID=$!

    MAX_SECS=2
    SECONDS_SINCE=0
    START_SPOOL=$(date -u +"%s")
    FILE_EXISTS=N

    while true; do

        if [[ -e ${SPOOL_TO} ]]; then
            FILE_EXISTS=Y
            debug "The requested spooled file (${SPOOL_TO}) was created. Nothing else to do"
            echo "SPOOL OFF" > ${FIFO_PATH} &
            break
        fi

        if [[ "${SECONDS_SINCE}" -ge "${MAX_SECS}" ]]; then
            debug "Didn't find requested spool file (${SPOOL_TO}) in ${SECONDS_SINCE} seconds. "
            kill ${SPOOL_PID} 2> /dev/null
            makeFifo
            exit ${TIMEOUT}
        fi

        sleep 0.1
        NOW_TIME=$(date -u +"%s")
        SECONDS_SINCE=$((NOW_TIME-START_SPOOL))

    done

    rm -f ${SPOOL_TO}

else
#Doesn't exist. Make the fifo and start sql
    debug "The FIFO did not exist on the system as yet"
    makeFifo
fi


lib/make.js
'use babel';
'use strict';

import path from 'path';
import fs from 'fs';
import tmp from 'tmp';

export function provideBuilder(){

    // Var to Keep track of the temporary file created when compiling against
    // an untitled file. Reason for this is that the `doMatch` function (which
    // returns a list of errors) doesn't have context to the build config.
    var _untitledFileDest;

    return class OracleBuildConfig {

        constructor(cwd){
            this.cwd = cwd;

            // Assume we have a saved file.
            // We need to keep track of this so that we can act in the pre and
            // post build phases to restore the configuration if necessary.
            this._savedFile = true;
        }

        getNiceName(){
            return 'Oracle Build Config';
        }

        isEligible(){
            this.oracleBuildConfig = path.join(this.cwd, '.atom-build-oracle.json');

            return fs.existsSync(this.oracleBuildConfig);
        }

        // preBuild: Before running the build, check out the active tab to see
        // if it is indeed a saved file or not. If it isn't, we want to save it
        // to a temporary file and point the build configuration to this new file
        preBuild(){

            // The only time the temp file will be opened is when there is an
            // issue in the compilation. Set an explanatory message that can
            // be added to the top of the untitled file contents when creating
            // the temporary file.
            const tmpFileWarning = [
                `-- This file was auto generated on ${new Date()} by`,
                `-- the \`build-oracle\` Atom plugin as the plugin requires a saved file to `,
                `-- compile and report any compilation errors. `,
                `-- `,
                `-- This file will be discarded upon exit from Atom; If you want to keep the `,
                `-- script you should save the file somewhere more practical.`
            ].join("\n");

            let fileBuffer = atom.workspace.getActivePaneItem();
            let isSavedFile = fileBuffer.buffer.file ? true : false;

            if (!isSavedFile){
                this._savedFile = false;
                // this._configInstance.savedFile = false;

                // the script being compiled is the last argument passed in to
                // the execute build command.
                const SQL_SCRIPT_INDEX = this.args.length-1;

                let untitledFile = tmp.fileSync({"postfix": ".sql"});
                _untitledFileDest = untitledFile.name;

                fs.writeFileSync(untitledFile.fd, `${tmpFileWarning}\n\n${fileBuffer.buffer.cachedText}`);

                // Set the build configuration to point to the created temporary file.
                this.args[SQL_SCRIPT_INDEX] = untitledFile.name;
                delete this.cwd;
            }
        }

        // postBuild: After the build has completed, we want to restore the
        // build configuration, to it's original state (which by default, compiles
        // the currently active file and the current working directory to the
        // path of the file).
        postBuild(buildOutcome, stdOut, stdErr){

            // If we operated on an untitled file, we need to restore
            // the configuration
            if (!this._savedFile) {
                // the script being compiled is the last argument passed in to
                // the execute build command.
                const SQL_SCRIPT_INDEX = this.args.length-1;

                this._savedFile = true;
                this.args[SQL_SCRIPT_INDEX] = '{FILE_ACTIVE}';
                this.cwd = '{FILE_ACTIVE_PATH}';
                _untitledFileDest = undefined;
            }

        }

        settings(){
            this.settingsFetched = new Date();
            var pkgDir = this.getPackagePath();
            let sqlPath = this.getPackageSetting('sqlPath');
            let nlsLang = this.getPackageSetting('nlsLang');
            let dyldLibraryPath = this.getPackageSetting('dyldLibraryPath');
            let enforceUtf8 = isWindows() ? this.getPackageSetting('enforceUtf8') : false;

            //OS X (El Capitan) needs this env variable set in order to find
            //the libraries to run SQL*Plus, and there is a security measure
            //which prevents it being passed into Atom.
            if (dyldLibraryPath){
                process.env.DYLD_LIBRARY_PATH = dyldLibraryPath;
            }

            if (!process.env.NLS_LANG && nlsLang){
                process.env.NLS_LANG = nlsLang;
            }

            //encforceUtf8: workaround for Windows is to set an environment variable
            //that the JVM will pick up so that UTF-8 encodings will work.
            //Only relevant for Windows.
            //See: http://stackoverflow.com/questions/361975/setting-the-default-java-character-encoding
            //https://github.com/tschf/atom-build-oracle/issues/29
            if (enforceUtf8){
                let encodingString = '-Dfile.encoding=UTF-8 ';
                if (process.env.JAVA_OPTS) {
                    encodingString += process.env.JAVA_OPTS;
                }

                process.env.JAVA_OPTS = encodingString;
            }

            var parsedConfig = this.getFileJson(this.oracleBuildConfig);
            var buildTargets = [];

            for (var i = 0; i < parsedConfig.length; i++){
                let currentTarget = parsedConfig[i];

                let user = currentTarget.user;
                let pass = currentTarget.password;
                let host = currentTarget.host;
                let port = currentTarget.port ? currentTarget.port.toString() : "1521";
                let sid = currentTarget.sid;
                let connectString = currentTarget.connectString;
                let thisConfig = {};

                if (!connectString){
                    connectString = `${user}/${pass}@//${host}:${port}/${sid}`;
                }
                thisConfig.name = 'Oracle Build::' + currentTarget.targetName;

                let compileScript = path.join(pkgDir, 'lib', 'compileSimple.sql');
                thisConfig.exec = sqlPath
                thisConfig.args = [
                  "-L",
                  connectString,
                  '@' + compileScript,
                  '{FILE_ACTIVE}'
                ];
                thisConfig.sh = false;
                thisConfig.cwd = "{FILE_ACTIVE_PATH}";
                thisConfig.functionMatch = this.doMatch;
                thisConfig.preBuild = this.preBuild;
                thisConfig.postBuild = this.postBuild;

                // custom attributes not a part of the provider
                // _savedFile: keep track if we are compiling on a saved file
                // or not. This will allow us to act in the pre and post build
                // changing the build config.
                thisConfig._savedFile = true;

                buildTargets.push(thisConfig);
            }

            return buildTargets;
        }

        doMatch(buildOutput){

            const buildOutputLines = buildOutput.split(/\n/);
            const errors = [];

            for(let i = 0; i < buildOutputLines.length; i++){
                const currentLine = buildOutputLines[i];
                const oraErrorMatches = currentLine.match(/((PLS|ORA)-[\d]+:.+$)/g);
                const dyldErrorMatches = currentLine.match(/dyld: Library not loaded:/g)
                const currentFile = _untitledFileDest ? _untitledFileDest : atom.workspace.getActivePaneItem().buffer.file.path;

                if (oraErrorMatches && oraErrorMatches.length > 0){
                    errors.push({
                        message: oraErrorMatches[0] + ' (Please review the build output for more detailed information)',
                        file: currentFile,
                        type: 'Error',
                        severity: 'error'
                    });
                } else if (dyldErrorMatches && dyldErrorMatches.length > 0){
                    errors.push({
                        message: 'DYLD_LIBRARY_PATH is not set. Try setting in the build-oracle settings.',
                        file: currentFile,
                        type: 'Error',
                        severity: 'error'
                    });
                }
            }

            return errors;
        }

        on(evt, evtCallback){
            //As as 25.11.2015, `on` only provides the `refresh` evt
            if (evt !== "refresh"){
                return;
            }

            this.fsWatcher = fs.watch(this.oracleBuildConfig, () => {
                var minTime = 1000;//1 second

                //Make sure a reasonable time has passed before calling the
                //refresh callback again
                if (new Date() - this.settingsFetched > minTime){
                    //calling `refreshTargets` in `atom-build/lib/build.js`
                    evtCallback();
                }
            });

            this.sqlPathObserver = atom.config.observe('build-oracle.sqlPath', (val) => {
                //first call, assign the current setting value
                if (!(this.sqlPath)){
                    this.sqlPath = val;
                } else {
                    //only if its changed should we `refreshTargets`
                    if (this.sqlPath !== val){
                        this.sqlPath = val;
                        evtCallback();
                    }
                }
            });

            this.nlsLangObserver = atom.config.observe('build-oracle.nlsLang', (val) => {
                //first call, assign the current setting value
                if (!(this.nlsLang)){
                    this.nlsLang = val;
                } else {
                    //only if its changed should we `refreshTargets`
                    if (this.nlsLang !== val){
                        this.nlsLang = val;
                        evtCallback();
                    }
                }
            });

            this.dyldLibraryPathObserver = atom.config.observe('build-oracle.dyldLibraryPath', (val) => {
                //first call, assign the current setting value
                if (!(this.dyldLibraryPath)){
                    this.dyldLibraryPath = val;
                } else {
                    //only if its changed should we `refreshTargets`
                    if (this.dyldLibraryPath !== val){
                        this.dyldLibraryPath = val;
                        evtCallback();
                    }
                }
            });

            this.enforceUtf8Observer = atom.config.observe('build-oracle.enforceUtf8', (val) => {
                //first call assign the current settings val
                if (!(this.hasOwnProperty('enforceUtf8'))){
                    this.enforceUtf8 = val;
                } else {
                    //only if it changes should we `refreshTargets`
                    if (this.enforceUtf8 !== val) {
                        this.enforceUtf8 = val;
                        evtCallback();
                    }

                }
            });
        }

        off(){
            //Called before a new refresh, so the existing fileWatch can be switched off
            if (this.fsWatcher){
                this.fsWatcher.close();
                this.fsWatcher = undefined;
            }

            if (this.sqlPathObserver){
                this.sqlPathObserver.dispose();
                this.sqlPathObserver = undefined;
            }

            if (this.nlsLangObserver){
                this.nlsLangObserver.dispose();
                this.nlsLangObserver = undefined;
            }

            if (this.dyldLibraryPathObserver) {
                this.dyldLibraryPathObserver.dispose();
                this.dyldLibraryPathObserver = undefined;
            }

            if (this.enforceUtf8Observer) {
                this.enforceUtf8Observer.dispose();
                this.enforceUtf8Observer = undefined;
            }
        }

        //Helper functions
        getPluginName(){
            return 'build-oracle';
        }

        getPackagePath(){
            var pkgManager = atom.packages;

            return pkgManager.resolvePackagePath(this.getPluginName());
        }

        getPackageSetting(propertyName){
            var pluginName = this.getPluginName();
            var settingValue = atom.config.get(`${pluginName}.${propertyName}`);

            return settingValue;
        }

        getFileJson(path){
            delete require.cache[path];

            return require(path);
        }
    }

};

function isWindows(){
    return /^win/.test(process.platform);
}

function isOsx(){
    return /^darwin$/.test(process.platform);
}

export function activate(){
    //Install package deps, see: https://github.com/steelbrain/package-deps
    require('atom-package-deps').install('build-oracle');
};

function config(){

    let buildConfig = {};
    let settingIndex = 0;

    buildConfig.sqlPath = {
        title: 'Path to SQL interpreter',
        type: 'string',
        default: 'sqlplus',
        description: 'Specify the path to SQL*Plus (or SQLcl). If its in your path, the basename is acceptable',
        order: ++settingIndex
    };

    buildConfig.nlsLang = {
        title: 'NLS_LANG',
        type: 'string',
        default: '.AL32UTF8',
        description: 'Specify the NLS_LANG to ensure the correct encoding is used. Format is: [NLS&#95;LANGUAGE]&#95;[NLS&#95;TERRITORY].[NLS&#95;CHARACTERSET].',
        order: ++settingIndex
    };

    if (isOsx()){
        buildConfig.dyldLibraryPath = {
          title: 'DYLD_LIBRARY_PATH',
          type: 'string',
          default: '',
          description: 'Specify the DYLD_LIBRARY_PATH so SQL*Plus can run properly',
          order: ++settingIndex
        }
    }

    //SQLcl on Windows doesn't seem to support UTF-8 encoding. Need a work around
    //in order to enable it.
    if (isWindows()){
        buildConfig.enforceUtf8 = {
            title: 'Enforce UTF-8 encoding',
            type: 'boolean',
            default: true,
            description: 'Enable UTF-8 encoding in the JVM (Used for SQLcl on Windows. Only works with sql.bat)',
            order: ++settingIndex
        }
    }

    return buildConfig;
}

module.exports.config = config();
